hazelcast-client/src/main/java/com/hazelcast/client/util/ClientDelegatingFuture.java

```diff
@@ -38,20 +39,14 @@
  */
 public class ClientDelegatingFuture<V> implements InternalCompletableFuture<V> {
 
+    private static final Object NULL_OBJECT = new Object();
     private final ClientInvocationFuture future;
     private final SerializationService serializationService;
     private final ClientMessageDecoder clientMessageDecoder;
     private final boolean deserializeResponse;
     private final V defaultValue;
-    private final Object mutex = new Object();
     private final Executor userExecutor;
-    private Throwable error;
-    private V deserializedValue;
-    /**
-     * mutex object is used as an initial NIL value
-     */
-    private volatile Object response = mutex;
-    private volatile boolean done;
+    private final AtomicReference<Object> responseMessage = new AtomicReference<Object>(NULL_OBJECT);
```

Comment:
please use a Atomicreferencefieldupdater. 

Using an  AtomicReference for every invocation will increase litter and this will increase gc pressure.